Improve accessibility of documentation by increasing contrast for different website elements

namurphy opened the pull request:
The default font color for in-line code samples in our docs has been light red.   Putting [a typical doc page in a web accessibility tool](https://wave.webaim.org/report#/https://docs.plasmapy.org/en/stable/particles/particle_class.html#creating-a-particle-object) shows that the font color does not meet the WCAG AA standard for contrast.  This PR is to change the color of in-line code samples so that it has sufficient contrast to meet this standard.  I'm also attempting to change it so that the `:file:` role leads to a different font color too.  

- [ ] Make a final decision on which colors to use for the different roles
- [ ] Add comments to the `.css` file that describes what each of the new content is doing.

I haven't found a good resource that describes custom CSS files with Sphinx, so I might add a little to the documentation guide about how to go about doing this too.

StanczakDominik commented:
So if I'm correct, the change influences code blocks such as [these](https://plasmapy--1287.org.readthedocs.build/en/1287/particles/particle_class.html#creating-a-particle-object), right?

rocco8773 commented:
> 
> 
> So if I'm correct, the change influences code blocks such as [these](https://plasmapy--1287.org.readthedocs.build/en/1287/particles/particle_class.html#creating-a-particle-object), right?

I'm going to say no since a code block is a `<div class="highlight"></div>` class whereas inline code is a `<code></code>` block.

rocco8773 commented:
So, I have no problem changing highlighting characteristics to improve readability, but I have to say the color choices here and in PR #1268 make things harder to ready.  While the color choices may have better contrast with the background, they bleed into the color of the surrounding text so they are not as "highlighted/distinguishable" as they were before. I mean the code color chosen here is dark blue which is almost identical to the black text next to it and, thus, does not pop that inline code.

namurphy commented:
My goal was to change the font color for the in-line literal text that we write in reST like: ``` ``x = 3`` ``` (← I needed ten back ticks to say this..) but it's also changing the `` `numpy.nan` ``-like things which do the Sphinx & intersphinx links to code objects.  (Only six back ticks this time!)

I still need to go back and fix this, but trying to understand CSS started making my brain hurt.

The color in the big code blocks is done using pygments, which I changed in #1268 to meet this accessibility standard.  (I tried looking into a Sphinx extension to get pygments formatting in in-line code snippets but wasn't able to get it to work.)

rocco8773 commented:
> 
> 
> My goal was to change the font color for the in-line literal text that we write in reST like: `` ``x = 3`` `` (← I needed ten back ticks to say this..) but it's also changing the `` `numpy.nan` ``-like things which do the Sphinx & intersphinx links to code objects. (Only six back ticks this time!)
> 
> I still need to go back and fix this, but trying to understand CSS started making my brain hurt.
> 
> The color in the big code blocks is done using pygments, which I changed in #1268 to meet this accessibility standard. (I tried looking into a Sphinx extension to get pygments formatting in in-line code snippets but wasn't able to get it to work.)

I get all that, my argument is that the chosen colors may improve contrast between the letters and the background but they make the text blend in with the surrounding text.  If syntax highlight blends in with the surrounding text, then that defeats the point of syntax highlighting.

I'll live with the code block highlighting since that will be a lot of work to fix.  But choosing a better single color for inline code blocks is not a difficult task to do.

namurphy commented:
Some notes for when I come back to this later... 

The html to get a link to `Particle` in our docs is...
```html
<code class = "xref py py-obj docutils literal notranslate">
  <span class="pre">Particle</span>
</code> 
```
The html to show the literal `"iron"` is...
```html
<code class="docutils literal notranslate">
  <span class="pre">"Fe"
</code>
```
The html to show a file path is:
```html
<code class="file docutils literal notranslate">
  <span class="pre">__init__.py</span>
</code>
```

StanczakDominik commented:
Doesn't this get superseded by #1361, which would let us close this one?

namurphy commented:
> Doesn't this get superseded by #1361, which would let us close this one?

No, unfortunately pygments only gets used for code blocks and not in-line text, at least.  These CSS customizations will be needed to style the literals that show up in-line rather than in a separate block.  

With that said, there might be a [Sphinx extension](https://sphinxcontrib-inlinesyntaxhighlight.readthedocs.io/en/latest/) that we could try instead, though that extension's [repo](https://github.com/sphinx-contrib/inlinesyntaxhighlight) hasn't been updated since 2013...

namurphy commented:
> While the color choices may have better contrast with the background, they bleed into the color of the surrounding text so they are not as "highlighted/distinguishable" as they were before.

This is a good point!  It turns out that there is an accessibility guideline for exactly this, which can be checked with a [link contrast checker](https://webaim.org/resources/linkcontrastchecker) with the "link to body text" check.  

I went through and updated the colors for the different elements, and added some CSS customizations from elsewhere that make the admonition color schemes more accessible too.  I think that the contrast levels now meet accessibility guidelines in the [WAVE accessibility check](https://wave.webaim.org/report#/https://plasmapy--1287.org.readthedocs.build/en/1287/) for the doc build from this pull request.  

With that said, I'm very open to different options for color scheme customizations, so long as the color choices meet accessibility guidelines.

rocco8773 commented on changelog/1287.doc.rst:
```diff
@@ -0,0 +1,3 @@
+Changed the color of in-line literals to be consistent with the
+`Web Content Accessibility Guidelines 2 Level AA Conformance
+<https://www.w3.org/WAI/WCAG2AA-Conformance>`__ for contrast.
```
Is there a specific section in the WCAG you can link to?  It would be good to link to the actual inline literal guidelines.

namurphy replied:
Done.

rocco8773 commented on .pre-commit-config.yaml:
```diff
@@ -54,3 +54,8 @@ repos:
     args: [--autofix]
   - id: pretty-format-yaml
     args: [--autofix, --indent, '2']
+
+- repo: https://github.com/pre-commit/mirrors-csslint
+  rev: v1.0.5
+  hooks:
+  - id: csslint
```
I'm thinking there should be a changelog entry that states this is being added.

namurphy replied:
Done.

rocco8773 reviewed (changes requested):
In addition to my comments, if we add `sphinx_rtd_theme_ext_color` to the `extensions` configuration variable in the `conf.py` then there should be a `doc` changelog entry noting it.

namurphy commented on docs/_static/rtd_theme_overrides.css:
```diff
@@ -71,14 +103,206 @@ div.pkgnav ul li {
 
 div.pkgnav ul li a:hover {
-    color: white;
+    color: #f50404;
```
Good idea!  I'll use one of those so long as they meet the access guidelines.

I propose we trademark `#ff2929` as PlasmaPy red. 😬  Or...maybe the open source equivalent of a trademark?

namurphy replied:
Neither of the colors met the color contrast guidelines, so I picked a color with the same RGB ratios as #ff2929 but were just enough to meet the guidelines.

namurphy commented on docs/_static/rtd_theme_overrides.css:
```diff
@@ -71,14 +103,206 @@ div.pkgnav ul li {
+    background: #277CB4;
+}
+
+/* error, danger */
+
+.rst-content .danger .admonition-title,
+.rst-content .danger .wy-alert-title,
+.rst-content .error .admonition-title,
+.rst-content .error .wy-alert-title,
+.rst-content .wy-alert-danger.admonition-todo .admonition-title,
+.rst-content .wy-alert-danger.admonition-todo .wy-alert-title,
+.rst-content .wy-alert-danger.admonition .admonition-title,
+.rst-content .wy-alert-danger.admonition .wy-alert-title,
+.rst-content .wy-alert-danger.attention .admonition-title,
+.rst-content .wy-alert-danger.attention .wy-alert-title,
+.rst-content .wy-alert-danger.caution .admonition-title,
+.rst-content .wy-alert-danger.caution .wy-alert-title,
+.rst-content .wy-alert-danger.hint .admonition-title,
+.rst-content .wy-alert-danger.hint .wy-alert-title,
+.rst-content .wy-alert-danger.important .admonition-title,
+.rst-content .wy-alert-danger.important .wy-alert-title,
+.rst-content .wy-alert-danger.note .admonition-title,
+.rst-content .wy-alert-danger.note .wy-alert-title,
+.rst-content .wy-alert-danger.seealso .admonition-title,
+.rst-content .wy-alert-danger.seealso .wy-alert-title,
+.rst-content .wy-alert-danger.tip .admonition-title,
+.rst-content .wy-alert-danger.tip .wy-alert-title,
+.rst-content .wy-alert-danger.warning .admonition-title,
+.rst-content .wy-alert-danger.warning .wy-alert-title,
+.rst-content .wy-alert.wy-alert-danger .admonition-title,
+.wy-alert.wy-alert-danger .rst-content .admonition-title,
+.wy-alert.wy-alert-danger .wy-alert-title {
+    background: #e31704;
+}
+
+/* Generic admonition titles */
+
+.wy-alert-title, .rst-content .admonition-title {
+    background: #277CB4;
+}
```
This only includes some of CSS from `sphinx_rtd_theme_ext_color` and not all of it, since I didn't want to override any of our other CSS customizations.  I think I also wanted to be able to customize the color schemes for admonitions to make sure there's still sufficient contrast with the hyperlink font colors.

I wondering if it'd be worth putting these customizations into a separate CSS file since they're (hopefully) temporary and they have a specific purpose.  Hm...

 - [ ] Make sure there's a note in here that these customizations should be dropped after the `sphinx_rtd_theme` defaults end up matching the WCAG guidelines.

rocco8773 replied:
I would test it out, but I'm fairly confident our CSS will still override `sphinx_rtd_theme_ext_color` if added to `extensions` since Sphinx places our stylesheet as the top priority.

rocco8773 replied:
After a little research, this is how I'd approach things...

1. create a new directory `docs/_static/css/` to house CSS files that modify the theme CSS file (i.e. that associated with `sphinx_rtd_theme`)
2. Move our CSS file `rtd_theme_overrides.css` into the newly crated directory and rename it `plasmapy.css`
3. Since `sphinx_rtd_theme_ext_color_contrast` is not released on PyPI, copy [their CSS file](https://github.com/AaltoSciComp/sphinx_rtd_theme_ext_color_contrast/blob/master/sphinx_rtd_theme_ext_color_contrast/_static/sphinx_rtd_theme_ext_color_contrast.css) into the newly created directory.  Obviously add a header to the file explaining and crediting where it came from.
4. Update the `setup()` function in our `conf.py` to add both CSS files using `app.add_css_file()`.  The [documentation for `add_css_file()`](https://www.sphinx-doc.org/en/master/extdev/appapi.html#sphinx.application.Sphinx.add_css_file) function show you can set priorities for the added CSS files; thus, we can ensure our CSS always takes priority.

namurphy replied:
I like!

namurphy replied:
Done, except that I still need to...
 - [ ] Re-check the documentation to make sure that everything is working as intended. 

I don't think there were any conflicts that would make if necessary for me to set CSS file priorities.

namurphy commented on docs/_static/rtd_theme_overrides.css:
```diff
@@ -29,7 +60,8 @@ div.graphviz {
  */
-.rst-content dl.rst table p, .rst-content dl.std table p,
+.rst-content dl.rst table p,
+.rst-content dl.std table p,l
```
Huh...I'm wondering if this came from a CSS autoformatter tool.  I'll have to follow up on this.

 - [ ] Check if this is a typo

rocco8773 replied:
I figured this was done by some autoformatter, but I've never seen it in a CSS.

namurphy replied:
This would likely be easier to figure out if I actually understood CSS... 😐

namurphy commented on docs/_static/rtd_theme_overrides.css:
```diff
@@ -71,14 +103,206 @@ div.pkgnav ul li {
+  https://webaim.org/resources/linkcontrastchecker
+*/
+
+div.wy-nav-content a:link { /* still changing links in sidebar!!!! */
```
- [ ] Make sure this comment is addressed in more recent changes I made, and if so, then remove it

namurphy commented:
pre-commit.ci autofix

namurphy commented:
Ok, I think it's ready for the next round of review.  Thanks!

rocco8773 commented on docs/_static/css/plasmapy.css:
```diff
@@ -0,0 +1,185 @@
+/*
+  This file contains styles rules to override styles from the Sphinx
+  Read the Docs theme. This file is used to further customize the style
+  of PlasmaPy's documentation.
```
Update to reflect this overrides RTD themes, as well as adds our own features.

namurphy commented on docs/_static/css/plasmapy.css:
```diff
@@ -0,0 +1,185 @@
+  rules. Right-click or ctrl-click on an element of a website, and
+  choose "inspect" to see the HTML code along with CSS selectors and
+  properties.
+
+  Alternatively, style rules for PlasmaPy's documentation can be viewed
+  at: https://docs.plasmapy.org/en/latest/_static/css/theme.css
```
```suggestion
```

rocco8773 commented on docs/_static/css/plasmapy.css:
```diff
@@ -0,0 +1,185 @@
+  choose "inspect" to see the HTML code along with CSS selectors and
+  properties.
+
+  Alternatively, style rules for PlasmaPy's documentation can be viewed
+  at: https://docs.plasmapy.org/en/latest/_static/css/theme.css
+
```
```suggestion
```

rocco8773 commented on docs/_static/css/plasmapy.css:
```diff
@@ -0,0 +1,185 @@
+  CSS files can be validated at: https://jigsaw.w3.org/css-validator
+*/
+
+/* Style for toctree in a table */
```
```suggestion
/* -----------------------------------------------------------------------------
 * RTD Overrides
 */
 
/* Style for toctree in a table */
```

Since this CSS can contain overrides and new styles, I think it helps readability and comprehensibility to point out what is overriding styles.

rocco8773 commented on docs/_static/css/plasmapy.css:
```diff
@@ -0,0 +1,182 @@
+}
+
+div.pkgnav ul {
+    /*margin: 3px 0px;*/
```
```suggestion
```

rocco8773 reviewed (approved):
Looks good to me! I don't know why RTD is not pulling the current CSS though. :/

namurphy commented:
Thank you for the reviews!  I'm still happy to adjust the styles later on if we want to.
